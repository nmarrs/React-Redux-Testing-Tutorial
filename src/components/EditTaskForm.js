import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './styles/EditTaskForm.css'

class EditTaskForm extends Component {
  static propTypes = {
    task: PropTypes.object,
    editTask: PropTypes.func.isRequired,
    toggleEditTaskModal: PropTypes.func.isRequired,
    isAdminView: PropTypes.bool.isRequired
  }

  state = {
    title: this.props.task.title,
    currentStatus: this.props.task.currentStatus,
    timeEstimate: this.props.task.timeEstimate,
    priority: this.props.task.priority,
    notes: this.props.task.notes,
    feedback: this.props.task.feedback
  }

  handleTitleInputChange = (event) => {
    this.setState({title: event.target.value})
  }

  handleStatusSelectChange = (event) => {
    this.setState({currentStatus: event.target.value})
  }

  handleTimeEstimateInputChange = (event) => {
    this.setState({timeEstimate: event.target.value})
  }

  handlePrioritySelectChange = (event) => {
    this.setState({priority: event.target.value})
  }

  handleNotesInputChange = (event) => {
    this.setState({notes: event.target.value})
  }

  handleFeedbackInputChange = (event) => {
    this.setState({feedback: event.target.value})
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const {
      title,
      timeEstimate,
      priority
    } = this.state

    const {
      task,
      isAdminView,
      editTask,
      toggleEditTaskModal
    } = this.props

    if (title === '' || timeEstimate === '0' || priority === '-') {
      window.alert('Error, task must have a valid title, time estimate, and priority level.')
      return
    }

    let editedTask = {
      ...task,
      ...this.state
    }

    editTask(editedTask, isAdminView)
    toggleEditTaskModal(false)
  }

  render() {
    const {
      title,
      currentStatus,
      timeEstimate,
      priority,
      notes,
      feedback
    } = this.state

    const {
      task
    } = this.props

    return (
      <div className='EditTaskForm-modal'>
        <h4 className='m-3'><i className="fas fa-info-circle"></i> Task Details</h4>
        <form>
          <div className='form-group'>
            <label htmlFor='taskTitle'><i className="fas fa-tag"></i> Title</label>
            <input className='form-control' id='taskTitleInput' defaultValue={task && task.title ? task.title : title} onChange={this.handleTitleInputChange} />
          </div>
          <div className="form-group">
            <label htmlFor='taskStatus'><i className="fas fa-bolt"></i> Status</label>
            <select className="form-control" id="taskStatusSelect" defaultValue={task && task.currentStatus ? task.currentStatus : currentStatus} onChange={this.handleStatusSelectChange}>
              <option>To Do</option>
              <option>In Progress</option>
              <option>Finished</option>
            </select>
          </div>
          <div className="form-group">
            <label htmlFor='taskPriority'><i className="fas fa-exclamation"></i> Priority Level (1-5)</label>
            <select className="form-control" id="taskPrioritySelect" defaultValue={task && task.priority ? task.priority : priority} onChange={this.handlePrioritySelectChange} disabled>
              <option>-</option>
              <option>1</option>
              <option>2</option>
              <option>3</option>
              <option>4</option>
              <option>5</option>
            </select>
          </div>
          <div className='form-group'>
            <label htmlFor='taskEstimate'><i className="far fa-clock"></i> Estimate (hours)</label>
            <input type='number' min='0' className='form-control' id='taskEstimateInput' defaultValue={task && task.timeEstimate ? task.timeEstimate : timeEstimate} onChange={this.handleTimeEstimateInputChange} disabled />
          </div>
          <div className='form-group'>
            <label htmlFor='taskNotes'><i className="far fa-sticky-note"></i> Notes</label>
            <textarea className='form-control EditTaskForm-textarea' id='taskNotesTextArea' defaultValue={task && task.notes ? task.notes : notes} onChange={this.handleNotesInputChange} />
          </div>
          <div className='form-group'>
            <label htmlFor='taskFeedback'><i className="far fa-comments"></i> Feedback</label>
            <textarea className='form-control EditTaskForm-textarea' id='taskFeedbackTextArea' defaultValue={task && task.feedback ? task.feedback : feedback} onChange={this.handleFeedbackInputChange} />
          </div>
          <button type='submit' className='btn btn-primary' onClick={this.handleSubmit}><i className="fas fa-check"></i> Submit</button>
        </form>
      </div>
    )
  }
}

export default EditTaskForm
