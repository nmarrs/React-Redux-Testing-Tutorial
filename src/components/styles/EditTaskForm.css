.EditTaskForm-modal {
  min-width: 350px;
}

.EditTaskForm-textarea {
  min-height: 100px;
}
