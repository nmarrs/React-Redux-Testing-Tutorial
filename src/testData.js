export let testTask = {
  id: 1,
  title: 'Unload merchandise',
  priority: 3,
  timeEstimate: 2,
  notes: 'Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 150
}

export let testTask1 = {
  id: 2,
  title: 'Stock shelves with unpacked items',
  priority: 1,
  timeEstimate: 10,
  notes: 'Take merchandise items and stock shelves with them. Be sure to place items in correct areas.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 10
}

export let testTask2 = {
  id: 3,
  title: 'Stock shelves with unpacked items',
  priority: 5,
  timeEstimate: 2,
  notes: 'Take merchandise items and stock shelves with them. Be sure to place items in correct areas.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 250
}
