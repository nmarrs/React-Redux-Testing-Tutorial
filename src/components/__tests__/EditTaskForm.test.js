import React from 'react'
import { shallow } from 'enzyme'

import EditTaskForm from '.././EditTaskForm'
import { testTask } from '../../testData'

describe('EditTaskForm', () => {
  let wrapper
  let editTask = jest.fn()
  let toggleEditTaskModal = jest.fn()

  beforeEach(() => {
    wrapper = shallow(<EditTaskForm
      task={testTask}
      editTask={editTask}
      toggleEditTaskModal={toggleEditTaskModal}
      isAdminView={true} />)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should exist', () => {
    expect(wrapper).toBeDefined()
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('initially should have expected state', () => {
    expect(wrapper.state()).toMatchSnapshot()
  })

  describe('handlers', () => {
    it('handleTitleInputChange', () => {
      let testEvent = {
        target: {
          value: 'New Title'
        }
      }

      wrapper.instance().handleTitleInputChange(testEvent)

      expect(wrapper.state().title).toEqual('New Title')
    })

    it('handleStatusSelectChange', () => {
      let testEvent = {
        target: {
          value: 'In Progress'
        }
      }

      wrapper.instance().handleStatusSelectChange(testEvent)

      expect(wrapper.state().currentStatus).toEqual('In Progress')
    })

    it('handleTimeEstimateInputChange', () => {
      let testEvent = {
        target: {
          value: '6'
        }
      }

      wrapper.instance().handleTimeEstimateInputChange(testEvent)

      expect(wrapper.state().timeEstimate).toEqual('6')
    })

    it('handlePrioritySelectChange', () => {
      let testEvent = {
        target: {
          value: '3'
        }
      }

      wrapper.instance().handlePrioritySelectChange(testEvent)

      expect(wrapper.state().priority).toEqual('3')
    })

    it('handleFeedbackInputChange', () => {
      let testEvent = {
        target: {
          value: 'Test Notes'
        }
      }

      wrapper.instance().handleNotesInputChange(testEvent)

      expect(wrapper.state().notes).toEqual('Test Notes')
    })

    it('handleFeedbackInputChange', () => {
      let testEvent = {
        target: {
          value: 'Test Feedback'
        }
      }

      wrapper.instance().handleFeedbackInputChange(testEvent)

      expect(wrapper.state().feedback).toEqual('Test Feedback')
    })

    describe('handleSubmit', () => {
      it('invalid data', () => {
        global.window.alert = jest.fn()
        let testEvent = {
          preventDefault: jest.fn()
        }
        wrapper.instance().setState({title: '', priority: '-', timeEstimate: '0'})

        wrapper.instance().handleSubmit(testEvent)

        expect(global.window.alert).toHaveBeenCalled()
        expect(global.window.alert).toHaveBeenCalledWith('Error, task must have a valid title, time estimate, and priority level.')
      })

      it('valid data', () => {
        let testEvent = {
          preventDefault: jest.fn()
        }

        wrapper.instance().handleSubmit(testEvent)

        expect(editTask).toHaveBeenCalled()
        expect(toggleEditTaskModal).toHaveBeenCalled()
        expect(toggleEditTaskModal).toHaveBeenCalledWith(false)
      })
    })
  })
})
