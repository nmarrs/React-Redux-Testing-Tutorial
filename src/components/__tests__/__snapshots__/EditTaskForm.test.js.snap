// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTaskForm initially should have expected state 1`] = `
Object {
  "currentStatus": "To Do",
  "feedback": "",
  "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
  "priority": 3,
  "timeEstimate": 2,
  "title": "Unload merchandise",
}
`;

exports[`EditTaskForm renders correctly 1`] = `
<div
  className="EditTaskForm-modal"
>
  <h4
    className="m-3"
  >
    <i
      className="fas fa-info-circle"
    />
     Task Details
  </h4>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskTitle"
      >
        <i
          className="fas fa-tag"
        />
         Title
      </label>
      <input
        className="form-control"
        defaultValue="Unload merchandise"
        id="taskTitleInput"
        onChange={[Function]}
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskStatus"
      >
        <i
          className="fas fa-bolt"
        />
         Status
      </label>
      <select
        className="form-control"
        defaultValue="To Do"
        id="taskStatusSelect"
        onChange={[Function]}
      >
        <option>
          To Do
        </option>
        <option>
          In Progress
        </option>
        <option>
          Finished
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskPriority"
      >
        <i
          className="fas fa-exclamation"
        />
         Priority Level (1-5)
      </label>
      <select
        className="form-control"
        defaultValue={3}
        disabled={true}
        id="taskPrioritySelect"
        onChange={[Function]}
      >
        <option>
          -
        </option>
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
        <option>
          4
        </option>
        <option>
          5
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskEstimate"
      >
        <i
          className="far fa-clock"
        />
         Estimate (hours)
      </label>
      <input
        className="form-control"
        defaultValue={2}
        disabled={true}
        id="taskEstimateInput"
        min="0"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskNotes"
      >
        <i
          className="far fa-sticky-note"
        />
         Notes
      </label>
      <textarea
        className="form-control EditTaskForm-textarea"
        defaultValue="Unload merchandise from the bay door. Driver usually comes by at 6 am sharp."
        id="taskNotesTextArea"
        onChange={[Function]}
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="taskFeedback"
      >
        <i
          className="far fa-comments"
        />
         Feedback
      </label>
      <textarea
        className="form-control EditTaskForm-textarea"
        defaultValue=""
        id="taskFeedbackTextArea"
        onChange={[Function]}
      />
    </div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="submit"
    >
      <i
        className="fas fa-check"
      />
       Submit
    </button>
  </form>
</div>
`;
